src/arch/aarch64/kernel/scheduler.rs

```diff
@@ -1,82 +1,253 @@
+		debug!("IST0 is located at {:#X}", ist0);*/
+		let stack = VirtAddr::zero();
+		let ist0 = VirtAddr::zero();
+
+		TaskStacks::Boot(BootStack { stack, ist0 })
+	}
+
+	pub fn get_user_stack_size(&self) -> usize {
+		match self {
+			TaskStacks::Boot(_) => 0,
+			TaskStacks::Common(stacks) => {
+				stacks.total_size - DEFAULT_STACK_SIZE - KERNEL_STACK_SIZE
+			}
+		}
+	}
+
+	pub fn get_user_stack(&self) -> VirtAddr {
+		match self {
+			TaskStacks::Boot(_) => VirtAddr::zero(),
+			TaskStacks::Common(stacks) => {
+				stacks.virt_addr + KERNEL_STACK_SIZE + DEFAULT_STACK_SIZE + 3 * BasePageSize::SIZE
+			}
+		}
+	}
+
+	pub fn get_kernel_stack(&self) -> VirtAddr {
+		match self {
+			TaskStacks::Boot(stacks) => stacks.stack,
+			TaskStacks::Common(stacks) => {
+				stacks.virt_addr + KERNEL_STACK_SIZE + 2 * BasePageSize::SIZE
+			}
+		}
+	}
+
+	pub fn get_kernel_stack_size(&self) -> usize {
+		match self {
+			TaskStacks::Boot(_) => KERNEL_STACK_SIZE,
+			TaskStacks::Common(_) => DEFAULT_STACK_SIZE,
+		}
+	}
+
+	pub fn get_interupt_stack(&self) -> VirtAddr {
+		match self {
+			TaskStacks::Boot(stacks) => stacks.ist0,
+			TaskStacks::Common(stacks) => stacks.virt_addr + BasePageSize::SIZE,
+		}
+	}
+
+	pub fn get_interupt_stack_size(&self) -> usize {
+		KERNEL_STACK_SIZE
+	}
+}
+
+impl Clone for TaskStacks {
+	fn clone(&self) -> TaskStacks {
+		match self {
+			TaskStacks::Boot(_) => TaskStacks::new(0),
+			TaskStacks::Common(stacks) => {
+				TaskStacks::new(stacks.total_size - DEFAULT_STACK_SIZE - KERNEL_STACK_SIZE)
+			}
 		}
 	}
 }
 
 impl Drop for TaskStacks {
 	fn drop(&mut self) {
-		if !self.is_boot_stack {
-			// TODO: Deallocate
+		// we should never deallocate a boot stack
+		match self {
+			TaskStacks::Boot(_) => {}
+			TaskStacks::Common(stacks) => {
+				debug!(
+					"Deallocating stacks at {:#X} with a size of {} KB",
+					stacks.virt_addr,
+					stacks.total_size >> 10,
+				);
+
+				crate::arch::mm::paging::unmap::<BasePageSize>(
+					stacks.virt_addr,
+					stacks.total_size / BasePageSize::SIZE + 4,
+				);
+				crate::arch::mm::virtualmem::deallocate(
+					stacks.virt_addr,
+					stacks.total_size + 4 * BasePageSize::SIZE,
+				);
+				crate::arch::mm::physicalmem::deallocate(stacks.phys_addr, stacks.total_size);
+			}
+		}
+	}
+}
+
+pub struct TaskTLS {
+	address: VirtAddr,
+	//fs: VirtAddr,
+	//layout: Layout,
+}
+
+impl TaskTLS {
+	pub fn new(tls_size: usize) -> Self {
+		Self {
+			address: VirtAddr::zero(),
 		}
 	}
+
+	#[inline]
+	pub fn address(&self) -> VirtAddr {
+		self.address
+	}
+}
+
+impl Drop for TaskTLS {
+	fn drop(&mut self) {
+		/*debug!(
+				"Deallocate TLS at 0x{:x} (layout {:?})",
+				self.address, self.layout,
+		);
+
+		unsafe {
+				dealloc(self.address.as_mut_ptr::<u8>(), self.layout);
+		}*/
+	}
+}
+
+impl Clone for TaskTLS {
+	fn clone(&self) -> Self {
+		TaskTLS::new(environment::get_tls_memsz())
+	}
 }
 
 extern "C" fn leave_task() -> ! {
-	core_scheduler().exit(0)
+	core_scheduler().exit(0);
```

Comment:
```suggestion
	core_scheduler().exit(0)
```

If the function is noreturn, no semilicolon should be added.

Reply:
Ok, I will fix it